program_transformers/src/token_extensions/mod.rs

```diff
@@ -0,0 +1,36 @@
+mod mint;
+mod token_account;
+
+use crate::{
+    error::{ProgramTransformerError, ProgramTransformerResult},
+    AccountInfo, DownloadMetadataNotifier,
+};
+use blockbuster::programs::token_extensions::TokenExtensionsProgramAccount;
+use sea_orm::DatabaseConnection;
+
+use self::{mint::handle_token2022_mint_account, token_account::handle_token2022_token_account};
+
+pub async fn handle_token_extensions_program_account<'a, 'b>(
+    account_update: &AccountInfo,
+    parsing_result: &'a TokenExtensionsProgramAccount,
+    db: &'b DatabaseConnection,
+    download_metadata_notifier: &DownloadMetadataNotifier,
+) -> ProgramTransformerResult<()> {
+    match &parsing_result {
```

Comment:
Does there need to be handling for burned tokens?

Reply:
In my review now. Getting an integration test in to confirm. This is a issue with token-metadata as well.

Reply:
Token Metadata has this handling if the metadata account is empty: https://github.com/metaplex-foundation/digital-asset-rpc-infrastructure/blob/main/program_transformers/src/token_metadata/mod.rs#L24-L27

But then I think how DAS effectively leaves out burned accounts is because the supply is zero, like here: https://github.com/metaplex-foundation/digital-asset-rpc-infrastructure/blob/main/digital_asset_types/src/dao/scopes/asset.rs#L67

So then supply is probably being set by `handle_token2022_mint_account` so maybe it is already not going to return them, but they would never be marked as burnt?

Reply:
@danenbm 

> Token Metadata has this handling if the metadata account is empty: https://github.com/metaplex-foundation/digital-asset-rpc-infrastructure/blob/main/program_transformers/src/token_metadata/mod.rs#L24-L27

Is the account owner not set to the system program so would this account update ever come through? Looking through this EmptyAccount creation though.

> But then I think how DAS effectively leaves out burned accounts is because the supply is zero, like here: https://github.com/metaplex-foundation/digital-asset-rpc-infrastructure/blob/main/digital_asset_types/src/dao/scopes/asset.rs#L67

This is based on token_account same issue occurs if token account is closed and ownership transferred system program.

Reply:
I've looked into EmptyAccount which is what triggers flagging an asset as burned.

It is set within the handle_account of blockbuster when the data slice for the account is empty.

This occurs in burn_nonfungible within the token-metadata program.

https://github.com/metaplex-foundation/blockbuster/blob/9b07c07dcdbbab468c0943c5b4d83916b99912c4/blockbuster/src/programs/token_metadata/mod.rs#L73-L78

https://github.com/metaplex-foundation/mpl-token-metadata/blob/main/programs/token-metadata/program/src/utils/mod.rs#L243-L51

The token metadata account is re-assigned to the system program so does geyser send a change event under the owner of token-metadata or system program? If its the later DAS will never register the change since its account selector does not include system program.


The token extensions blockbuster parser does include the same EmptyAccount but again depends on the ownership assignment.

Reply:
On a local test validator stack when a traditional NFT is created and then the burn instruction of token-metadata is invoked. The following occurs:

1. The mint is left open but supply dropped to zero. This is correctly reflected in DAS.
2. The token account is closed and reassigned to the system program but no DAS is not notified.
3. token metadata account is left open with 1 byte. DAS receives the update event but throws 
```
{"timestamp":"2024-04-09T13:15:37.272578Z","level":"WARN","fields":{"message":"Error parsing account 5Yga8VnpA1GXnYohtH7w3NMKVczA7ZXUUA16T5UJEG16: \"Uninitialized account type\"","log.target":"nft_ingester::metrics","log.module_path":"nft_ingester::metrics","log.file":"nft_ingester/src/metrics.rs","log.line":94},"target":"nft_ingester::metrics"}
```
The asset is not marked as burnt because it the update event fails to parse.

Reply:
This is because account closures go through the system program. It's annoying. We'll upstream this fix.